File: java/leetcode4.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        if (nums1.length < nums2.length) {
            return findMedianSortedArrays(nums2, nums1);
        }

        int m = nums1.length, n = nums2.length;

        int halfLen = (m + n + 1) >> 1;
        int jleft = 0, jright = n;
        int i = 0, j = 0;
        while(jleft <= jright) {
            j = (jleft + jright) >> 1;
            i = halfLen - j;
            if (j < n && nums2[j] < nums1[i-1]) jleft = j + 1;
            else if(j > 0 && nums1[i] < nums2[j-1]) jright =  j - 1;
            else break;
        }

        double maxLeft, minRight;
        if (j == 0) maxLeft = nums1[i-1];
        else if(j == n && i==0) maxLeft = nums2[j-1];
        else maxLeft = Math.max(nums1[i-1], nums2[j-1]);

        if ((m + n) % 2 == 1) return maxLeft;

        if (j == n) minRight = nums1[i];
        else if(j == 0 && i == m) minRight = nums2[j];
        else minRight = Math.min(nums1[i], nums2[j]);
    
        return (maxLeft + minRight) / 2;
    }
}

File: python/leetcode4.py
class Solution:
    def findMedianSortedArrays(self, nums1, nums2):
        m, n = len(nums1), len(nums2)
        if m < n:
            m, n, nums1, nums2 = n, m, nums2, nums1
        
        half_len = (m + n + 1) >> 1
        jleft, jright = 0, n
        while jleft <= jright:
            j = (jleft + jright) >> 1
            i = half_len - j
            if j < n and nums2[j] < nums1[i-1]:
                jleft = j + 1
            elif j > 0 and nums1[i] < nums2[j-1]:
                jright = j - 1
            else:
                break
        
        if j == 0:
            max_left = nums1[i-1]
        elif j == n and i == 0:
            max_left = nums2[j-1]
        else:
            max_left = max(nums1[i-1], nums2[j-1])
        
        if (m + n) % 2 == 1:
            return max_left
        
        if j == n:
            min_right = nums1[i]
        elif j == 0 and i == m:
            min_right = nums2[j]
        else:
            min_right = min(nums1[i], nums2[j])
        
        return (max_left + min_right) / 2
